pkg/runtime/bulk_subscriber.go

```diff
@@ -0,0 +1,706 @@
+		bulkSubDiag.statusWiseDiag[string(pubsub.Retry)] -= int64(len(data))
+	}
+	req := &pubsub.BulkPublishRequest{
+		Entries:    data,
+		PubsubName: name,
+		Topic:      deadLetterTopic,
+		Metadata:   msg.Metadata,
+	}
+
+	_, err := a.BulkPublish(req)
+	if err != nil {
+		log.Errorf("error sending message to dead letter, origin topic: %s dead letter topic %s err: %w", msg.Topic, deadLetterTopic, err)
+	}
+	return err
+}
+
+func validateEntryID(entryID string, i int) error {
+	if entryID == "" {
+		log.Warn("Invalid blank entry id received while processing bulk pub/sub event, won't be able to process it")
+		return errors.New("Blank entryID supplied - won't be able to process it")
+	}
+	return nil
+}
+
+func populateBulkSubcribedMessage(message *pubsub.BulkMessageEntry, event interface{},
+	routePathBulkMessageMap *map[string]pubsubBulkSubscribedMessage,
+	rPath string, i int, msg *pubsub.BulkMessage, isCloudEvent bool, psName string, contentType string,
+) {
+	childMessage := runtimePubsub.BulkSubscribeMessageItem{
+		Event:       event,
+		Metadata:    message.Metadata,
+		EntryID:     message.EntryID,
+		ContentType: contentType,
+	}
+	var cloudEvent map[string]interface{}
+	mapTypeEvent, ok := event.(map[string]interface{})
+	if ok {
+		cloudEvent = mapTypeEvent
+	}
+	if val, ok := (*routePathBulkMessageMap)[rPath]; ok {
+		if isCloudEvent {
+			val.cloudEvents[val.length] = mapTypeEvent
+		}
+		val.rawData[val.length] = childMessage
+		val.entries[val.length] = &msg.Entries[i]
+		val.length++
+		(*routePathBulkMessageMap)[rPath] = val
+	} else {
+		cloudEvents := make([]map[string]interface{}, len(msg.Entries))
+		rawDataItems := make([]runtimePubsub.BulkSubscribeMessageItem, len(msg.Entries))
+		rawDataItems[0] = childMessage
+		entries := make([]*pubsub.BulkMessageEntry, len(msg.Entries))
+		entries[0] = &msg.Entries[i]
+		if isCloudEvent {
+			cloudEvents[0] = cloudEvent
+		}
+		psm := pubsubBulkSubscribedMessage{
+			cloudEvents: cloudEvents,
+			rawData:     rawDataItems,
+			entries:     entries,
+			topic:       msg.Topic,
+			metadata:    msg.Metadata,
+			pubsub:      psName,
+			length:      1,
+		}
+		(*routePathBulkMessageMap)[rPath] = psm
+	}
+}
+
+func populateBulkSubscribeResponsesWithError(entries []*pubsub.BulkMessageEntry,
+	bulkResponses *[]pubsub.BulkSubscribeResponseEntry, entryIDIndexMap *map[string]int, err error,
+) {
+	for _, item := range entries {
+		ind := (*entryIDIndexMap)[item.EntryID]
+		if (*bulkResponses)[ind].EntryID == "" {
+			(*bulkResponses)[ind].EntryID = item.EntryID
+			(*bulkResponses)[ind].Error = err
+		}
+	}
+}
+
+func populateAllBulkResponsesWithError(bulkMsg *pubsub.BulkMessage,
+	bulkResponses *[]pubsub.BulkSubscribeResponseEntry, err error,
+) {
+	for i, item := range bulkMsg.Entries {
+		if (*bulkResponses)[i].EntryID == "" {
+			(*bulkResponses)[i].EntryID = item.EntryID
+			(*bulkResponses)[i].Error = err
+		}
+	}
+}
+
+func newBulkSubIngressDiagnostics() bulkSubIngressDiagnostics {
```

Comment:
Reviewers: Initializes a map with different status possible. These counters will be incremented per message response.